libs/vm/include/vm/state.hpp

```diff
@@ -216,13 +223,15 @@ class State : public IState
     return v;
   }
 
-  void FlushIO()
+  Status FlushIO()
   {
     // if we have an IO observer then inform it of the changes
     if (!vm_->HasError() && vm_->HasIoObserver())
     {
-      WriteHelper(name_, value_, vm_->GetIOObserver());
+      return WriteHelper(name_, value_, vm_->GetIOObserver());
     }
+
+    return Status::OK;
```

Comment:
Is the status OK if it was not able to write?

Reply:
Have been thinking this through a few times before, and it felt like this is the right implementation if we are going by return value only (= no exception).
Reason being that this (`State::OK`) is returned only under conditions when:
 * VM already is in error scenario (`vm_->HasError() == true`), in which case it does not make much sense to go ahead report another nested error 
 * and/or when VM does not have registered IO observer, which is responsible for actual write to storage (`vm_->HasIoObserver() == false`), what was, by original design intention, considered as valid state (whoever is setting up VM can decide not to register IO observer) - see original usage of the observer in the `State::State(...)` constructor.

However, I have to admit that original design intention somehow never felt completely kosher exactly for the same reason which caught you eye.

If we want to change this, there are 2 ways to go:
 1) either throw exception here in the case when `vm_->HasIoObserver() == false`, 
 2) Check IO observer validity at the `State<...>` construction time and fail at that point. Keep in mind that VM has `VM::SetIOObserver(IoObserverInterface &observer)` method, which allows to change IO observer instance - it can not be invalidated (null-ed), just changed to another object (what is however highly improbable scenario in real life)